### PR TITLE
Byron protocol info cleanups

### PR DESCRIPTION
**Pull request by deepfire**
Deduce more PBFT parameters implicit in genesis, instead of having them passed as demo configuration.

There is likely a discussion to be had, where to place the new functions and types.

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo.hs [hunk 1]
```diff
@@ -30,4 +30,5 @@ protocolInfo nodes nid demoProtocol = case demoProtocol of
     ProtocolMockPraos      params          -> protocolInfoPraos     nodes nid params
     ProtocolLeaderSchedule params schedule -> protocolInfoPraosRule nodes nid params schedule
     ProtocolMockPBFT       params          -> protocolInfoMockPBFT  nodes nid params
-    ProtocolRealPBFT       params gc mplc  -> protocolInfoByron     nodes nid params gc mplc
+    ProtocolRealPBFT       gc mplc mthr ->
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo.hs]: Just keep it on one line, like the others above it.

Comment (deepfire) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo.hs]: Done!

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs [hunk 2]
```diff
@@ -36,28 +41,64 @@ import           Ouroboros.Consensus.Ledger.Byron.Config
...
+-- | Given a delegation certificate, assign it a PBFT id, in the context of a fixed delegation map.
+--   The order is defined by the delegator's KeyHash ordering.
+certificateNodeIdPbft :: Cardano.Genesis.GenesisDelegation -> Cardano.Delegation.Certificate -> Maybe CoreNodeId
+certificateNodeIdPbft Cardano.Genesis.UnsafeGenesisDelegation{Cardano.Genesis.unGenesisDelegation=dlg} cert =
+  CoreNodeId <$> List.findIndex ((== cert) . snd) (Map.toAscList dlg)
+
+-- | Given a delegate key, assign it a PBFT id, in the context of a fixed delegation map.
+--   The order is defined by the delegator's KeyHash ordering.
+delegateNodeIdPbft :: Cardano.Genesis.GenesisDelegation -> Cardano.VerificationKey -> Maybe CoreNodeId
+delegateNodeIdPbft dlg vk = join $ certificateNodeIdPbft dlg <$> findCertByDelegateByron dlg vk
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: I'm pretty sure this is the wrong algorithm. The right algorithm was in my original WIP patch.

What we should do is this:
* take the supplied delegation cert, grab the genesis key (hash) from that
* look up that genesis key (hash) in the set of known genesis key hashes (from the genesis file) and if present, determine the index.

```
+    genesisKeyHashes :: Set Cardano.Common.KeyHash
+    genesisKeyHashes = Cardano.Genesis.unGenesisKeyHashes
+                     . Cardano.Genesis.configGenesisKeyHashes
+                     $ genesisConfig
+
+    -- The CoreNodeId index is determined by the 0-based position in the
+    -- sort order of the genesis key hashes.
+    -- TODO: this is important policy, it should live in the core PBFT module.
+    genesisKeyCoreNodeId :: Cardano.ProtocolMagicId
+                         -> VerKeyDSIGN CardanoDSIGN
+                         -> Maybe CoreNodeId
+    genesisKeyCoreNodeId magic vkey =
+        Data.Reflection.give magic $
+        CoreNodeId <$> Set.lookupIndex (hashVerKey vkey) genesisKeyHashes
```

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs [hunk 3]
```diff
@@ -79,7 +120,7 @@ protocolInfoByron (NumCoreNodes numCoreNodes) nid params gc mLeader =
               , pbftGenesisConfig   = gc
               , pbftGenesisHash     = coerce Cardano.Genesis.configGenesisHeaderHash gc
               , pbftEpochSlots      = Cardano.Genesis.configEpochSlots gc
-              , pbftGenesisDlg      = Cardano.Genesis.configHeavyDelegation gc
+              , pbftGenesisDlg      = delegation
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: I wouldn't bother changing this, it's clearer how it was. Achieving sharing here doesn't buy anything.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs [hunk 4]
```diff
@@ -96,13 +137,19 @@ protocolInfoByron (NumCoreNodes numCoreNodes) nid params gc mLeader =
...
+    delegation :: Cardano.Genesis.GenesisDelegation
+    delegation = Cardano.Genesis.configHeavyDelegation gc
+    proofOfCredentials :: Cardano.Genesis.GenesisDelegation -> PbftLeaderCredentials -> PBftIsLeader PBftCardanoCrypto
+    proofOfCredentials dlg cred@(PbftLeaderCredentials sk cert) =
       PBftIsLeader
-      { pbftCoreNodeId = nid' -- TODO (Duncan): compute this based on the genesis verification key
+      { pbftCoreNodeId = if checkDelegationSubject vk cert
+                         then flip fromMaybe (delegateNodeIdPbft dlg vk) $
+                              throw $ ByronInconsistentLeaderCredentials cred
+                         else throw $ ByronInconsistentLeaderCredentials cred
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: In principle we should do this check in an earlier context, and not throw pure errors for what is a validation check, not a programmer error.

I'd suggest pulling this out as a top level exported verification function, and using it when resolving the config in the application. If you do that, you can still make it a check here, like an assertion, but then you have to document it as a pre-condition and provide that way of the app level doing the check first.

Here's a nice way to do it: don't export the `PbftLeaderCredentials` constructor but export a function that makes one, and it can do the check and return `Either ByronInconsistentLeaderCredentials PbftLeaderCredentials`. Though actually that adds zero extra information, so just use `Maybe PbftLeaderCredentials` and then you don't need `ByronInconsistentLeaderCredentials` at all.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs [hunk 5]
```diff
@@ -96,13 +137,19 @@ protocolInfoByron (NumCoreNodes numCoreNodes) nid params gc mLeader =
       , pInfoInitState  = ()
       }
   where
-    proofOfCredentials :: CoreNodeId -> PbftLeaderCredentials -> PBftIsLeader PBftCardanoCrypto
-    proofOfCredentials nid' (PbftLeaderCredentials sk dlg) =
+    delegation :: Cardano.Genesis.GenesisDelegation
+    delegation = Cardano.Genesis.configHeavyDelegation gc
+    proofOfCredentials :: Cardano.Genesis.GenesisDelegation -> PbftLeaderCredentials -> PBftIsLeader PBftCardanoCrypto
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: Let's stick to 80 cols.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs [hunk 6]
```diff
@@ -69,7 +110,7 @@ protocolInfoByron (NumCoreNodes numCoreNodes) nid params gc mLeader =
                       -- TODO: make pbftSignatureThreshold default to a value computed from pbftNumNodes,
                       --       but configurable?
                     }
-                , pbftIsLeader = proofOfCredentials nid <$> mLeader
+                , pbftIsLeader = proofOfCredentials delegation <$> mLeader
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: We don't need to pass this, it's already available in scope.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs [hunk 7]
```diff
@@ -96,13 +137,19 @@ protocolInfoByron (NumCoreNodes numCoreNodes) nid params gc mLeader =
       , pInfoInitState  = ()
       }
   where
-    proofOfCredentials :: CoreNodeId -> PbftLeaderCredentials -> PBftIsLeader PBftCardanoCrypto
-    proofOfCredentials nid' (PbftLeaderCredentials sk dlg) =
+    delegation :: Cardano.Genesis.GenesisDelegation
+    delegation = Cardano.Genesis.configHeavyDelegation gc
+    proofOfCredentials :: Cardano.Genesis.GenesisDelegation -> PbftLeaderCredentials -> PBftIsLeader PBftCardanoCrypto
+    proofOfCredentials dlg cred@(PbftLeaderCredentials sk cert) =
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: Once you fix the algorithm for `genesisKeyCoreNodeId` then you will not need this first arg at all.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs [hunk 8]
```diff
@@ -139,6 +138,8 @@ protocolInfoByron (NumCoreNodes numCoreNodes) params gc mLeader =
   where
     delegation :: Cardano.Genesis.GenesisDelegation
     delegation = Cardano.Genesis.configHeavyDelegation gc
+    numCoreNodes = Set.size . Cardano.Genesis.unGenesisKeyHashes $
+                   Cardano.Genesis.configGenesisKeyHashes gc
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: :+1:  Though note that we'll need the `genesisKeyHashes` as a `Set` on it's own for the `genesisKeyCoreNodeId` impl. So this will become
```
numCoreNodes = Set.size genesisKeyHashes
```

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs [hunk 9]
```diff
@@ -92,25 +95,34 @@ delegateNodeIdPbft dlg vk = join $ certificateNodeIdPbft dlg <$> findCertByDeleg
 checkDelegationSubject :: Cardano.VerificationKey -> Cardano.Delegation.Certificate -> Bool
 checkDelegationSubject vk = (== vk) . Cardano.Delegation.delegateVK
 
+-- | See chapter 4.1 of
+--   https://hydra.iohk.io/job/Cardano/cardano-ledger-specs/byronChainSpec/latest/download-by-type/doc-pdf/blockchain-spec
+mainnetPBftSignatureThreshold :: PBftSignatureThreshold
+mainnetPBftSignatureThreshold = PBftSignatureThreshold 7
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: Wait, what? Where did this come from? It's 0.22 isn't it?

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs [hunk 10]
```diff
@@ -92,25 +95,34 @@ delegateNodeIdPbft dlg vk = join $ certificateNodeIdPbft dlg <$> findCertByDeleg
...
+
+-- | Derive production-capable PBFT params.
+computePbftParams :: Cardano.Genesis.GenesisData -> Maybe PBftSignatureThreshold -> PBftParams
+computePbftParams Cardano.Genesis.GenesisData{Cardano.Genesis.gdGenesisKeyHashes, Cardano.Genesis.gdK=Cardano.Common.BlockCount k} mThreshold =
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: Let's stick to 80 cols.

File: ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT.hs [hunk 11]
```diff
@@ -132,7 +133,7 @@ data PBftParams = PBftParams {
 
       -- | Signature threshold. This represents the proportion of blocks in a
       -- pbftSignatureWindow-sized window which may be signed by any single key.
-    , pbftSignatureThreshold :: Double
+    , pbftSignatureThreshold :: PBftSignatureThreshold
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT.hs]: This is an unrelated change in core consensus code. Leave it out of this PR.

File: ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT.hs [hunk 12]
```diff
@@ -216,11 +221,11 @@ instance (PBftCrypto c, Typeable c) => OuroborosTag (PBft c) where
             $ throwError (PBftExceededSignThreshold totalSigners gkSigners)
           return $! takeR (winSize + 2*k) chainState Seq.|> (gk, blockSlot b)
     where
-      PBftParams{..} = pbftParams
+      PBftParams{pbftSignatureThreshold=PBftSignatureThreshold sigThresh, ..} = pbftParams
       PBftFields{..} = headerPBftFields cfg b
       winSize = fromIntegral pbftSignatureWindow
       SecurityParam (fromIntegral -> k) = pbftSecurityParam
-      wt = floor $ pbftSignatureThreshold * fromIntegral winSize
+      wt = floor $ sigThresh * fromIntegral winSize
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT.hs]: Let's not mess with this in a PR about configuration plumbing. This is critical consensus code.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs [hunk 13]
```diff
@@ -98,11 +98,16 @@ checkDelegationSubject vk = (== vk) . Cardano.Delegation.delegateVK
 -- | See chapter 4.1 of
 --   https://hydra.iohk.io/job/Cardano/cardano-ledger-specs/byronChainSpec/latest/download-by-type/doc-pdf/blockchain-spec
 mainnetPBftSignatureThreshold :: PBftSignatureThreshold
-mainnetPBftSignatureThreshold = PBftSignatureThreshold 7
+mainnetPBftSignatureThreshold = PBftSignatureThreshold 0.22
```
Comment (dcoutts) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: :relieved:

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs (hunk 9 above)
Comment (deepfire) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: I fixed that in a later commit!

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs (hunk 4 above)
Comment (deepfire) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: @dcoutts, this raises a philosophical question in my head -- if we're not exporting the dumb constructor from `ProtocolInfo.Byron`, are we ok with an expression that assumes the consistency ensured by the smart constructor?

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs (hunk 8 above)
Comment (deepfire) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: But I've factored `genesisKeyCoreNodeId` into PBFT, as per TODO..

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs (hunk 13 above)
Comment (deepfire) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: I didn't expect that people do review commit-by-commit -- and that was an early typo : -)

File: ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT.hs (hunk 12 above)
Comment (deepfire) [reply ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT.hs]: Done!

File: ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT.hs (hunk 11 above)
Comment (deepfire) [reply ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT.hs]: Done!

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs (hunk 6 above)
Comment (deepfire) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: Done.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs (hunk 5 above)
Comment (deepfire) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: Done.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs (hunk 3 above)
Comment (deepfire) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: Done.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs (hunk 2 above)
Comment (deepfire) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: Done.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs (hunk 4 above)
Comment (dcoutts) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: Yes, smart constructors that establish invariants is a standard pattern.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs (hunk 7 above)
Comment (deepfire) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: Done.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs (hunk 10 above)
Comment (deepfire) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: Done.

**Comment by deepfire** [comment]
https://github.com/input-output-hk/cardano-node/pull/79 is now using this -- https://github.com/input-output-hk/ouroboros-network/pull/824/commits/bc9fd760c08b7d7ce0665ddb9233b2cabdb68564, specifically.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo.hs [hunk 14]
```diff
@@ -26,8 +26,8 @@ protocolInfo :: NumCoreNodes
...
-    ProtocolMockBFT        params          -> protocolInfoBft       nodes nid params
-    ProtocolMockPraos      params          -> protocolInfoPraos     nodes nid params
-    ProtocolLeaderSchedule params schedule -> protocolInfoPraosRule nodes nid params schedule
-    ProtocolMockPBFT       params          -> protocolInfoMockPBFT  nodes nid params
-    ProtocolRealPBFT       params gc mplc  -> protocolInfoByron     nodes nid params gc mplc
+    ProtocolMockBFT        params               -> protocolInfoBft       nodes nid params
+    ProtocolMockPraos      params               -> protocolInfoPraos     nodes nid params
+    ProtocolLeaderSchedule params schedule      -> protocolInfoPraosRule nodes nid params schedule
+    ProtocolMockPBFT       params               -> protocolInfoMockPBFT  nodes nid params
+    ProtocolRealPBFT       gc mthr prv swv mplc -> protocolInfoByron     gc mthr prv swv mplc
```
Comment (nc6) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo.hs]: Bikeshedding, but changes like this are why I don't like styles which align identifiers. There's only one change here, but it looks like a whole block has changed.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs [hunk 15]
```diff
@@ -40,46 +48,113 @@ import qualified Test.Cardano.Chain.Genesis.Dummy as Dummy
...
-                    }
-                , pbftIsLeader = proofOfCredentials nid <$> mLeader
-                }
+            encNodeConfigP   = PBftNodeConfig {
+                pbftParams          =  PBftParams
+                  { pbftSecurityParam      = SecurityParam (fromIntegral kParam)
+                  , pbftNumNodes           = fromIntegral . Set.size
+                                           . Genesis.unGenesisKeyHashes
+                                           $ genesisKeyHashes
+                  , pbftSignatureWindow    = fromIntegral kParam
```
Comment (nc6) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: Are we mandating that the signing window is always `k` long? If so, should we add this to the specs?

Comment (dcoutts) [reply ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: We probably should yes. The `cardano-sl` code interprets PBFT in such a way as it's always a K sized window. The spec leaves it as a free param. Our analysis (the 0.22) was based on the window being K. So yes I think we should fix it to K in the specs, and not make it a independent param in the impl.

**Comment by dcoutts** [comment]
bors +r

**Comment by dcoutts** [comment]
Yes of course I did! :roll_eyes:

File: ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs [hunk 16]
```diff
@@ -40,46 +48,113 @@ import qualified Test.Cardano.Chain.Genesis.Dummy as Dummy
...
+
+    return PBftLeaderCredentials {
+      plcSignKey     = sk
+    , plcDlgCert     = cert
+    , plcCodeNodeId  = nid
     }
+  where
+    (?!) :: Maybe a -> e -> Either e a
+    Just x  ?! _ = Right x
+    Nothing ?! e = Left  e
```
Comment (deepfire) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node/ProtocolInfo/Byron.hs]: @dcoutts, should we add this to the `cardano-prelude` (or otherwise similar place)?
